numba_dpex/caching.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2022 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+
+from numba.core.caching import IndexDataCacheFile, _Cache, _CacheImpl
+from numba.core.serialize import dumps
+
+from numba_dpex.core import compiler
+
+
+class DpexCacheImpl(_CacheImpl):
+    def reduce(self, cres):
+        """
+        Returns a serialized CompileResult
+        """
+        return cres._reduce()
+
+    def rebuild(self, target_context, payload):
+        """
+        Returns the unserialized CompileResult
+        """
+        return compiler.CompileResult._rebuild(target_context, *payload)
+
+    def check_cachable(self, cres):
+        # For the time being, assuming all numba-dpex Kernels are always cachable.
+        return True
+
+
+class DpexCache(_Cache):
+    """
+    Implements a cache that saves and loads CUDA kernels and compile results.
+    """
+
+    _impl_class = DpexCacheImpl
+
+    def __init__(self, py_func):
+        self._name = repr(py_func)
+        self._py_func = py_func
+        self._impl = self._impl_class(py_func)
+        self._cache_path = self._impl.locator.get_cache_path()
+        # This may be a bit strict but avoids us maintaining a magic number
+        source_stamp = self._impl.locator.get_source_stamp()
+        filename_base = self._impl.filename_base
+        self._cache_file = IndexDataCacheFile(
+            cache_path=self._cache_path,
+            filename_base=filename_base,
+            source_stamp=source_stamp,
+        )
+        self.enable()
+
+    @property
+    def cache_path(self):
+        return self._cache_path
+
+    def enable(self):
+        self._enabled = True
+
+    def disable(self):
+        self._enabled = False
+
+    def flush(self):
+        self._cache_file.flush()
+
+    def load_overload(self, sig, target_context):
+        if not self._enabled:
+            return
+        # key = self._index_key(sig, target_context.codegen())
+        key = self._index_key(sig)
+        data = self._cache_file.load(key)
+        if data is not None:
+            data = self._impl.rebuild(target_context, data)
+        print("-----> caching.load_overload()")
+        return data
+
+    def save_overload(self, sig, data):
+        if not self._enabled:
+            return
+        if not self._impl.check_cachable(data):
+            return
+        self._impl.locator.ensure_cache_path()
+        # key = self._index_key(sig, data.codegen)
```

Comment:
@diptorupd This `data` is a compiled object from `spirc_kernel.py` ([here](https://github.com/IntelPython/numba-dpex/pull/815/files#diff-3d7cb4cb29c1041552d90ddd499fe5b9fd08b817c9495772924d3e225660cfb0R120)). But the compiled kernel doesn't have `codegen`.